client.go

```diff
@@ -25,9 +25,62 @@ type Client struct {
 	BaseURL      string
 	StreamingURL string
 	hc           *http.Client
+	token        *oauth2.Token
+}
+
+// WithToken provides an oauth2.Token to the client for auth.
+func WithToken(t *oauth2.Token) func(c *Client) error {
+	return func(c *Client) error {
+		c.token = t
+		return nil
+	}
+}
+
+// WithTokenFile reads a JSON serialized oauth2.Token struct from disk and provides it
+// to the client for auth.
+func WithTokenFile(path string) func(c *Client) error {
+	t, err := loadToken(path)
+	if err != nil {
+		return func(c *Client) error {
+			return err
+		}
+	}
+	return WithToken(t)
+}
+
+// New creates a new Tesla API client. You must provided one of WithToken or WithTokenFile
+// functional options to initialize the client with an OAuth token.
+func New(ctx context.Context, options ...func(*Client) error) (*Client, error) {
```

Comment:
I wonder if having `func(*Client) error` have a defined type would make this cleaner.

Maybe something like
```
type ClientOption func(c *Client) error
```

Then we'd have
```
func WithToken(t *oauth2.Token) ClientOption {
   ...
}
func WithTokenFile(path string) ClientOption {
   ...
}
func New(ctx context.Context, options ...ClientOption) (*Client, error) {
   ...
}
```

Thoughts?

Reply:
I like it.